### PR TITLE
Add links for primitives in "jump to definition" feature

### PR DESCRIPTION
**Pull request by GuillaumeGomez**
Follow-up of #84176.

I created a function `primitive_from_str` which is code that was originally in `collect_intra_doc_links::resolve_primitive` to prevent code duplication.

I also created the `primitive_link_url` function which is somewhat similar to `primitive_link` but too much different to merge both of them.

Part of https://github.com/rust-lang/rust/issues/89095.

r? @jyn514

### CODE REVIEW

File: src/librustdoc/html/format.rs [hunk 1]
```diff
@@ -676,6 +676,36 @@ fn primitive_link(
...
 
+crate fn primitive_link_url(
+    prim: clean::PrimitiveType,
+    cx: &Context<'_>,
+    root_path: &str,
+) -> Option<String> {
+    let m = &cx.cache();
+    match m.primitive_locations.get(&prim) {
+        Some(&def_id) if def_id.is_local() => {
+            Some(format!("{}primitive.{}.html", root_path, prim.as_sym(),))
```
Comment (jyn514) [inline src/librustdoc/html/format.rs:687]: Shouldn't this include the crate name? Are there any tests that the links this generates aren't giving 404s?

File: src/librustdoc/html/format.rs [hunk 2]
```diff
@@ -676,6 +676,36 @@ fn primitive_link(
...
+        }
+        Some(&def_id) => {
+            let path = match m.extern_locations[&def_id.krate] {
+                ExternalLocation::Remote(ref s) => {
+                    format!("{}/std", s.trim_end_matches('/'))
```
Comment (jyn514) [inline src/librustdoc/html/format.rs:692]: Hard-coding `/std` definitely seems wrong. This should use `tcx.crate_name(def_id.krate)` instead.

File: src/librustdoc/html/format.rs [hunk 3]
```diff
@@ -676,6 +676,36 @@ fn primitive_link(
     Ok(())
 }
 
+crate fn primitive_link_url(
+    prim: clean::PrimitiveType,
+    cx: &Context<'_>,
+    root_path: &str,
+) -> Option<String> {
+    let m = &cx.cache();
+    match m.primitive_locations.get(&prim) {
```
Comment (jyn514) [inline src/librustdoc/html/format.rs:685]: You can make this a little smaller by using `m.primitive_locations.get()?` instead of matching.

File: src/librustdoc/html/format.rs (hunk 1 above)
Comment (GuillaumeGomez) [reply src/librustdoc/html/format.rs:687]: Yes it should actually... I'll add one.

File: src/librustdoc/html/format.rs (hunk 3 above)
Comment (GuillaumeGomez) [reply src/librustdoc/html/format.rs:685]: Good point!

File: src/librustdoc/html/format.rs [hunk 4]
```diff
@@ -676,6 +676,25 @@ fn primitive_link(
...
+    if def_id.is_local() {
+        return Some(format!("{}{}/primitive.{}.html", root_path, crate_name, prim.as_sym()));
+    }
+    let path = match m.extern_locations[&def_id.krate] {
+        ExternalLocation::Remote(ref s) => format!("{}/{}", s.trim_end_matches('/'), crate_name),
+        ExternalLocation::Local => format!("{}{}", root_path, crate_name),
+        ExternalLocation::Unknown => format!("{}/{}", crate::DOC_RUST_LANG_ORG_CHANNEL, crate_name),
+    };
+    Some(format!("{}/primitive.{}.html", path, prim.as_sym(),))
+}
```
Comment (jyn514) [inline src/librustdoc/html/format.rs:696]: Hmm, I kind of want to wait for https://github.com/rust-lang/rust/pull/87073 so we can just delete this function altogether. Would you be ok with that? I hope to get it merged within the week.

**Review by jyn514 — COMMENTED** [review]
I think this may be fixed without having to add all this extra work once #87073 is merged - do you mind waiting for that? You could also try rebasing over that in the meantime to see if the tests pass without your changes.

File: src/librustdoc/html/format.rs (hunk 1 above)
Comment (jyn514) [reply src/librustdoc/html/format.rs]: Is there any way to test the link is now correct? I'm worried this will get out of sync with other parts of rustdoc.

**Comment by GuillaumeGomez** [comment]
> I think this may be fixed without having to add all this extra work once #87073 is merged - do you mind waiting for that? You could also try rebasing over that in the meantime to see if the tests pass without your changes.

I'm totally fine with waiting for it!

File: src/librustdoc/html/format.rs (hunk 1 above)
Comment (GuillaumeGomez) [reply src/librustdoc/html/format.rs]: I added a test with a "local" primitive to check the full URL.

**Comment by JohnCSimon** [comment]
triage: fyi  #87073 is still open

**Comment by GuillaumeGomez** [comment]
#87073 has been merged, the PR has been rebased and updated to take the changes into account making it much smaller. Ready for the next review round! :)

File: src/librustdoc/html/render/span_map.rs [hunk 5]
```diff
@@ -73,6 +74,22 @@ impl<'tcx> SpanMapVisitor<'tcx> {
                 Some(def_id)
             }
             Res::Local(_) => None,
+            Res::PrimTy(p) => {
+                let span = path_span.unwrap_or_else(|| path.span);
+                if let Some(prim) = self
+                    .tcx
+                    .sess
+                    .source_map()
+                    .span_to_snippet(span)
```
Comment (jyn514) [inline src/librustdoc/html/render/span_map.rs:83]: Please do not use span_to_snippet, it's not reliable (e.g. it will break for libstd if people use --remap-path-prefix). Why not use PrimitiveType::from unconditionally?

File: src/librustdoc/html/render/span_map.rs [hunk 6]
```diff
@@ -73,6 +74,22 @@ impl<'tcx> SpanMapVisitor<'tcx> {
                 Some(def_id)
             }
             Res::Local(_) => None,
+            Res::PrimTy(p) => {
```
Comment (jyn514) [inline src/librustdoc/html/render/span_map.rs:77]: Note that not all primitives show up in PrimTy, it can only represent "path-like" primitives (so not e.g. arrays and tuples). Is that intentional?
https://doc.rust-lang.org/nightly/nightly-rustc/rustc_hir/hir/enum.PrimTy.html

File: src/test/rustdoc/check-source-code-urls-to-def.rs [hunk 7]
```diff
@@ -40,5 +42,9 @@ pub fn foo(a: u32, b: &str, c: String, d: Foo, e: bar::Bar, f: source_code::Sour
 
 // @has - '//a[@href="../../src/foo/auxiliary/source-code-bar.rs.html#14-16"]' 'bar::sub::Trait'
 // @has - '//a[@href="../../src/foo/auxiliary/source-code-bar.rs.html#14-16"]' 'Trait'
-pub fn foo2<T: bar::sub::Trait, V: Trait>(t: &T, v: &V) {
+pub fn foo2<T: bar::sub::Trait, V: Trait>(t: &T, v: &V, b: bool) {
 }
+
+// @has - '//a[@href="../../foo/primitive.bool.html"]' 'bool'
+#[doc(primitive = "bool")]
+mod whatever {}
```
Comment (jyn514) [inline src/test/rustdoc/check-source-code-urls-to-def.rs:50]: This seems unnecessarily complicated - can you check it links to doc.rust-lang.org instead, so we're testing how it behaves when there's no doc(primitive) available? That should be the case for literally every crate other than libstd and core.

Feel free to add a separate test for that if you want to keep this one, but I do think it's important to have.

**Review by jyn514 — CHANGES_REQUESTED** [review]

**Comment by jyn514** [comment]
You're right, this is a lot simpler now :)

File: src/test/rustdoc/check-source-code-urls-to-def.rs (hunk 7 above)
Comment (GuillaumeGomez) [reply src/test/rustdoc/check-source-code-urls-to-def.rs:50]: I think I'll add a second test so both cases are covered.

File: src/librustdoc/html/render/span_map.rs (hunk 5 above)
Comment (GuillaumeGomez) [reply src/librustdoc/html/render/span_map.rs:83]: I don't remember the reason so let's just use it unconditionnally.

File: src/librustdoc/html/render/span_map.rs (hunk 6 above)
Comment (GuillaumeGomez) [reply src/librustdoc/html/render/span_map.rs:77]: Yes it is. However a FIXME would be nice, I'll add one.

File: src/librustdoc/clean/utils.rs [hunk 8]
```diff
@@ -600,3 +600,36 @@ pub(super) fn display_macro_source(
         }
     }
 }
+
+crate fn primitive_from_str(path_str: &str) -> Option<PrimitiveType> {
```
Comment (jyn514) [inline src/librustdoc/clean/utils.rs:604]: This is no longer used anywhere except intra-doc links, right? Can you revert this change? I don't think most places should be using strings to parse primitives.

File: src/librustdoc/html/highlight.rs [hunk 9]
```diff
@@ -584,6 +585,13 @@ fn string<T: Display>(
                             .ok()
                             .map(|(url, _, _)| url)
                     }
+                    LinkFromSrc::Primitive(prim) => format::href_with_root_path(
+                        *PrimitiveType::primitive_locations(context.tcx()).get(&prim).unwrap(),
```
Comment (jyn514) [inline src/librustdoc/html/highlight.rs:589]:
```suggestion
                        PrimitiveType::primitive_locations(context.tcx())[&prim],
```

File: src/librustdoc/html/render/span_map.rs [hunk 10]
```diff
@@ -73,6 +74,12 @@ impl<'tcx> SpanMapVisitor<'tcx> {
...
+                // FIXME: Doesn't handle "path-like" primitives like arrays or tuples.
+                let span = path_span.unwrap_or_else(|| path.span);
```
Comment (jyn514) [inline src/librustdoc/html/render/span_map.rs:79]:
```suggestion
                let span = path_span.unwrap_or(path.span);
```

File: src/test/rustdoc/check-source-code-urls-to-def-std.rs [hunk 11]
```diff
@@ -0,0 +1,17 @@
...
+
+#![crate_name = "foo"]
+
+// @has 'src/foo/check-source-code-urls-to-def-std.rs.html'
+
+fn babar() {}
+
+// @has - '//a[@href="https://doc.rust-lang.org/nightly/std/primitive.u32.html"]' 'u32'
+// @has - '//a[@href="https://doc.rust-lang.org/nightly/std/primitive.str.html"]' 'str'
+// @has - '//a[@href="https://doc.rust-lang.org/nightly/std/primitive.bool.html"]' 'bool'
```
Comment (jyn514) [inline src/test/rustdoc/check-source-code-urls-to-def-std.rs:11]: These will break on beta.
```suggestion
// @has - '//a[@href="{{channel}}/std/primitive.u32.html"]' 'u32'
// @has - '//a[@href="{{channel}}/std/primitive.str.html"]' 'str'
// @has - '//a[@href="{{channel}}/nightly/std/primitive.bool.html"]' 'bool'
```

Comment (GuillaumeGomez) [reply src/test/rustdoc/check-source-code-urls-to-def-std.rs:11]: I didn't even know about `{{channel}}`, thanks!

File: src/librustdoc/clean/utils.rs (hunk 8 above)
Comment (GuillaumeGomez) [reply src/librustdoc/clean/utils.rs:604]: Ah right, moving it back then!

File: src/test/rustdoc/check-source-code-urls-to-def-std.rs (hunk 11 above)
Comment (jyn514) [reply src/test/rustdoc/check-source-code-urls-to-def-std.rs:11]: I just added it a few months ago: https://github.com/rust-lang/rust/pull/85990

**Comment by GuillaumeGomez** [comment]
Updated!

**Comment by jyn514** [comment]
r=me once CI is passing
